sway-core/src/semantic_analysis/storage_only_types.rs

```diff
@@ -0,0 +1,256 @@
+                (),
+                warnings,
+                errors
+            );
+            check!(expr_validate(expr), (), warnings, errors)
+        }
+        TypedDeclaration::FunctionDeclaration(TypedFunctionDeclaration {
+            body,
+            parameters,
+            ..
+        }) => {
+            check!(
+                validate_decls_for_storage_only_types_in_codeblock(body),
+                (),
+                warnings,
+                errors
+            );
+            for param in parameters {
+                check!(
+                    check_type(param.type_id, param.type_span.clone(), false),
+                    continue,
+                    warnings,
+                    errors
+                );
+            }
+        }
+        TypedDeclaration::AbiDeclaration(TypedAbiDeclaration { methods: _, .. })
+        | TypedDeclaration::TraitDeclaration(TypedTraitDeclaration { methods: _, .. }) => {
+            // These methods are not typed. They are however handled from ImplTrait.
+        }
+        TypedDeclaration::ImplTrait(TypedImplTrait { methods, .. }) => {
+            for method in methods {
+                check!(
+                    validate_decls_for_storage_only_types_in_codeblock(&method.body),
+                    continue,
+                    warnings,
+                    errors
+                )
+            }
+        }
+        TypedDeclaration::StructDeclaration(TypedStructDeclaration { fields, .. }) => {
+            for field in fields {
+                check!(
+                    check_type(field.type_id, field.span.clone(), false),
+                    continue,
+                    warnings,
+                    errors
+                );
+            }
+        }
+        TypedDeclaration::EnumDeclaration(TypedEnumDeclaration { variants, .. }) => {
+            for variant in variants {
+                check!(
+                    check_type(variant.type_id, variant.span.clone(), false),
+                    continue,
+                    warnings,
+                    errors
+                );
+            }
+        }
+        TypedDeclaration::StorageDeclaration(TypedStorageDeclaration { fields, .. }) => {
+            for field in fields {
+                check!(
+                    check_type(field.type_id, field.name.span().clone(), true),
+                    continue,
+                    warnings,
+                    errors
+                );
+            }
+        }
+        TypedDeclaration::GenericTypeForFunctionScope { .. }
+        | TypedDeclaration::ErrorRecovery
+        | TypedDeclaration::StorageReassignment(_)
+        | TypedDeclaration::Break { .. }
+        | TypedDeclaration::Continue { .. } => {}
+    }
+    ok((), warnings, errors)
+}
+
+pub fn validate_decls_for_storage_only_types_in_ast(
+    ast_n: &TypedAstNodeContent,
+) -> CompileResult<()> {
+    ast_node_validate(ast_n)
+}
```

Comment:
This is a bit redundant isn't it?  `ast_node_validate()` isn't `pub` so there's a slight differentiator but they might as well be one function?

Reply:
Yes, kind of. I just wanted to make the entry point explicit, since the private functions are mutually recursive.